### PR TITLE
feat(route-viewer-overlay): support clipping route outside of flex zones

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
This PR adds a prop to the `route-viewer-overlay` that allows not drawing a route within that route's flex zones. 

Looking for feedback on: 
- How the points are processed. It does a bad job when the points are badly defined, or right next to the zone. A library might do this better.

### CODE REVIEW

**Comment by binh-dam-ibigroup** [comment]
On first impression, I would try to show the difference between when the `clipToPatternStops` is applied or not, either using separate stories or using story args that provide interactive control of that parameter. As a nice-to-have, perhaps render the flex stops as part of the story/stories to provide the visual reference of where the clipping is supposed to happen.

File: packages/route-viewer-overlay/src/index.js [hunk 1]
```diff
@@ -55,13 +56,33 @@ class RouteViewerOverlay extends MapLayer {
     Object.values(routeData.patterns).forEach(pattern => {
       if (!pattern.geometry) return;
       const pts = polyline.decode(pattern.geometry.points);
+      let clippedPts = pts;
+      if (clipToPatternStops) {
+        // First, go through all stops to find flex zones
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/index.js:61]: This block could be a function.

Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.js]: Good call! Done in b8c1753

**Comment by miles-grant-ibigroup** [comment]
I originally wanted to add `StopsOverlay` to the story to show how the clipping works, but ran into too many issues with the import (as `StopsOverlay` isn't a dependency of the `RouteViewerOverlay` package). I have, however, added a button to enable and disable the clipping. 

If we'd really like to show the zone we could add a raw `GeoJSON` leaflet layer and render the stops into there... Let me know what you think.

**Comment by binh-dam-ibigroup** [comment]
> If we'd really like to show the zone we could add a raw `GeoJSON` leaflet layer and render the stops into there... Let me know what you think.

That would be ideal!

**Comment by miles-grant-ibigroup** [comment]
GeoJson layer is added in 80d872c!

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js [hunk 2]
```diff
@@ -32,3 +36,28 @@ export const WithPathStyling = () => (
...
+
+// Storyshot can't render a story with hooks. Creating a function like this
+// is the workaround. See https://github.com/storybookjs/storybook/issues/8177#issuecomment-599866282
+const FlexRouteWithClipButton = () => {
+  const [clip, setClip] = useState(true);
+  return (
+    <>
+      <button type="button" onClick={() => setClip(!clip)}>
+        {clip ? "Unclip" : "Clip"} Route to Outside Flex Zone
+      </button>
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/RouteViewerOverlay.story.js:48]: Replace this button with "story args", see https://storybook.js.org/blog/introducing-storybook-args/ and examples in #323 or #281.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js [hunk 3]
```diff
@@ -32,3 +36,28 @@ export const WithPathStyling = () => (
...
+      <BaseMap center={POWDER_SPRINGS} zoom={zoom}>
+        <RouteViewerOverlay
+          clipToPatternStops={clip}
+          routeData={flexRouteData}
+          visible
+        />
+        {/* Since the data is fixed, we know that stops[1] will contain the relevant flex zone. 
+        Using the stopsOverlay is not possible as it is very complex to implement */}
+        <GeoJSON data={flexRouteData.stops[1].geometries.geoJson} />
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/RouteViewerOverlay.story.js:57]: +1

File: packages/route-viewer-overlay/src/index.js [hunk 4]
```diff
@@ -16,6 +17,28 @@ const isGeomComplete = routeData => {
   );
 };
 
+// helper function that removes all points from array of points that are
+// within flex zones defined in an array of stops
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/index.js:21]: Use JSDoc style `/**` comment.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Use Storybook's Story Args to let users set the clipping prop and it will be good to go.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js (hunk 2 above)
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/RouteViewerOverlay.story.js]: I've moved to args and added a control in 6945aafe0033c070e7556d47894b2c7de63b4441, but worry that it's a bit less discoverable than the button. Moving to template also has required a few hacks

File: packages/route-viewer-overlay/src/index.js (hunk 4 above)
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.js]: Done in 6945aafe0033c070e7556d47894b2c7de63b4441

**Comment by miles-grant-ibigroup** [comment]
Clipping prop can now be set via Storybook Args, but I worry that it's more difficult to use and the code more messy. Let me know what you think.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js [hunk 5]
```diff
@@ -1,34 +1,76 @@
...
   title: "RouteViewerOverlay",
-  component: RouteViewerOverlay
+  component: RouteViewerOverlay,
+  argTypes: {
+    center: hiddenProp,
+    zoom: hiddenProp,
+    routeData: hiddenProp,
+    path: hiddenProp,
+    extraLayer: hiddenProp
+  }
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/RouteViewerOverlay.story.js:30]: Sort attributes.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js [hunk 6]
```diff
@@ -1,34 +1,76 @@
+/* eslint-disable react/jsx-props-no-spreading */
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/RouteViewerOverlay.story.js:1]: Neah, the props were previously set explicitly for `<BaseMap>` and `<RouteViewerOverlay>`.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js [hunk 7]
```diff
@@ -1,34 +1,76 @@
...
-      routeData={routeData}
-      visible
-    />
-  </BaseMap>
-);
+export const Default = Template.bind({});
+Default.args = {
+  center: PORTLAND,
+  zoom,
+  routeData
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/src/RouteViewerOverlay.story.js:49]: Sort attributes (also in stories below).

**Review by binh-dam-ibigroup — APPROVED** [review]
Besides some finishing touches it's good to go.

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js (hunk 5 above)
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/RouteViewerOverlay.story.js]: Sorted in f13469bc7ec8b24af48c993101ec3ca3bb69afcb

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js (hunk 6 above)
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/RouteViewerOverlay.story.js]: Oh good call don't know why I didn't think of this -- cleaned up in f13469bc7ec8b24af48c993101ec3ca3bb69afcb

File: packages/route-viewer-overlay/src/RouteViewerOverlay.story.js (hunk 7 above)
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/RouteViewerOverlay.story.js]: Sorted in f13469bc7ec8b24af48c993101ec3ca3bb69afcb

**Comment by miles-grant-ibigroup** [comment]
Thanks for that prop spreading catch! Looking much better.

**Review by philip-cline — APPROVED** [review]

**Pull request merged**